gridcomps/History/Sampler/MAPL_StationSamplerMod.F90

```diff
@@ -332,74 +484,196 @@ subroutine append_file(this,current_time,rc)
-             call this%formatter%put_var(xname,p_dst_2d,&
-                  start=[1,this%obs_written],count=[this%nstation,1],_RC)
-          end if
-          call ESMF_FieldDestroy(dst_field,nogarbage=.true.)
-       else if (rank==3) then
-          call ESMF_FieldGet(src_field,farrayptr=p_src_3d,_RC)
-          call ESMF_FieldGet(src_field,ungriddedLBound=lb,ungriddedUBound=ub,_RC)
-          if (this%vdata%lm/=(ub(1)-lb(1)+1)) then
-             lb(1)=1
-             ub(1)=this%vdata%lm
-          end if
-          dst_field = ESMF_FieldCreate(this%esmf_ls,name=xname,&
-               typekind=ESMF_TYPEKIND_R4,ungriddedLBound=lb,ungriddedUBound=ub,_RC)
-          call ESMF_FieldGet(dst_field,farrayptr=p_dst_3d,_RC)
-          call this%regridder%regrid(p_src_3d,p_dst_3d,_RC)
-          if (mapl_am_i_root()) then
-             nx=size(p_dst_3d,1); nz=size(p_dst_3d,2); allocate(arr(nz, nx), _STAT)
-             arr=reshape(p_dst_3d,[nz,nx],order=[2,1])
-             call this%formatter%put_var(xname,arr,&
-                  start=[1,1,this%obs_written],count=[nz,nx,1],_RC)
-             !note:     lev,station,time
-             deallocate(arr)
+    lm = this%vdata%lm
+    field_ds_2d = ESMF_FieldCreate (this%LS_ds, name='field_2d_ds', typekind=ESMF_TYPEKIND_R4, _RC)
+    field_chunk_2d = ESMF_FieldCreate (this%LS_chunk, name='field_2d_chunk', typekind=ESMF_TYPEKIND_R4, _RC)
+    dst_field = ESMF_FieldCreate (this%LS_ds, name='dst_field', typekind=ESMF_TYPEKIND_R4, &
+         gridToFieldMap=[1],ungriddedLBound=[1],ungriddedUBound=[lm],_RC)
+
+
+    !   caution about zero-sized array for MPI
+    !   redist
+    nx_sum = this%nstation
+    call ESMF_VMGetCurrent(vm,_RC)
+    call ESMF_VMGet(vm, mpiCommunicator=mpic, petCount=petCount, localPet=mypet, _RC)
+
+    iroot = 0
+    ip = mypet
+    N = nx_sum
+    M = petCount
+    nsend = int(ip+1, INT64) * int(N, INT64) / int(M, INT64) - &
+         int(ip  , INT64) * int(N, INT64) / int(M, INT64)
+    allocate ( recvcount (petCount) )
+    allocate ( displs    (petCount) )
+    do ip=0, M-1
+       recvcount(ip+1) =  int(ip+1, INT64) * int(N, INT64) / int(M, INT64) - &
+            int(ip  , INT64) * int(N, INT64) / int(M, INT64)
+    end do
+    displs(1)=0
+    do i = 2, petCount
+       displs(i) = displs(i-1) + recvcount(i-1)
+    end do
+
+    nsend_v = nsend * lm      ! vertical
+    allocate (recvcount_v, source = recvcount * lm )
+    allocate (displs_v, source = displs * lm )
+
+    if (mapl_am_i_root()) then
+       allocate ( p_rt_2d(nx_sum) )
+    else
+       allocate ( p_rt_2d(1) )
+    end if
+
+    ! p_rt_3d (lm, nx)
+    if (mapl_am_i_root()) then
+       allocate ( p_rt_3d(lm, nx_sum) )
+       allocate ( p_rt_3d_aux(nx_sum, lm) )
+    else
+       allocate ( p_rt_3d(lm, 1) )
+       allocate ( p_rt_3d_aux(1,lm) )
+    end if
+
+
+    iter = this%items%begin()
+    do while (iter /= this%items%end())
+       item => iter%get()
+       if (item%itemType == ItemTypeScalar) then
```

Comment:
Overly long nested conditional starting here.   Please consider how it could be refactored.  Can possibly work on it with you paired-programming later this week.

Reply:
These do while and if conditions shared the same layout as in GriddedIO.F90